contracts/tokenomics/vesting/src/contract.rs

```diff
@@ -320,26 +326,134 @@ fn compute_available_amount(current_time: u64, vesting_info: &VestingInfo) -> St
+                end_point.amount.checked_sub(schedule.start_point.amount)?,
+                time_period,
+            );
+            available_amount = available_amount.checked_add(release_amount)?;
+        }
+    }
+
+    Ok(available_amount)
+}
+
+/// Withdraw tokens from active vesting schedule.
+///
+/// Withdraw is possible if there is only one active vesting schedule.
+/// Active schedule's remaining amount must be greater than withdraw amount.
+/// This function terminates current active schedule (updates end_point)
+/// and creates a new one with remaining amount minus withdrawn amount.
+/// Withdrawn amount receiver is either `receiver` or `info.sender`.
+fn withdraw_from_active_schedule(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    account: String,
+    receiver: Option<String>,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if info.sender != config.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let acc = deps.api.addr_validate(&account)?;
+    let mut vesting_info = VESTING_INFO.load(deps.storage, &acc)?;
+    let block_time = env.block.time.seconds();
+
+    let mut active_schedules = vesting_info.schedules.iter_mut().filter(|schedule| {
+        if let Some(end_point) = schedule.end_point {
+            block_time >= schedule.start_point.time && block_time < end_point.time
+        } else {
+            block_time >= schedule.start_point.time
+        }
+    });
+
+    let new_schedule;
+    if let Some(schedule) = active_schedules.next() {
+        // Withdraw is not allowed if there are multiple active schedules
+        if active_schedules.next().is_some() {
+            return Err(ContractError::MultipleActiveSchedules(account));
+        }
+
+        let sch_unlocked_amount = calc_schedule_unlocked_amount(schedule, block_time)?;
+
+        let amount_left = schedule
+            .end_point
+            .map(|point| point.amount)
+            .unwrap_or(schedule.start_point.amount)
+            .checked_sub(sch_unlocked_amount)?;
+        if amount >= amount_left {
+            return Err(ContractError::NotEnoughTokens(amount_left));
+        }
+
+        new_schedule = if let Some(end_point) = &schedule.end_point {
+            VestingSchedule {
+                start_point: VestingSchedulePoint {
+                    time: block_time + 1,
+                    amount: Uint128::zero(),
+                },
+                end_point: Some(VestingSchedulePoint {
+                    time: end_point.time,
+                    amount: end_point.amount - sch_unlocked_amount - amount,
+                }),
+            }
+        } else {
+            VestingSchedule {
+                start_point: VestingSchedulePoint {
+                    time: block_time + 1,
+                    amount: schedule.start_point.amount - sch_unlocked_amount - amount,
```

Comment:
User could withdraw starting amount at this point. Let's exclude one-time schedules

Reply:
True. Fixed